Add Loss of Separation to usepe_logger.py

joakimrydberg opened the pull request:
The log for Loss of Separation is separate from the one for Conflicts, but is structured in the same way.

Considering the analysis that is to come, are there any objections to logging each parameter separately? Would you rather have them in one log file with an additional field denoting the type of event (conflict, LoS, etc...)?

@dulli I have used list comprehension for this variant, including the two lists `startlos` and `endlos`. Is this how you envisioned it? I couldn't find a way to sort the pairs of `currentlos` without a one-line for loop, as `sort()` works directly on the list rather than return a sorted list.

Depending on your comments I may modify the code for logging conflicts as well, before merge.

joakimrydberg committed:
Implement the logger for loss of separation

Creates a separate log for loss of separation

joakimrydberg committed:
Modify los_logger() to use list comprehension

dulli commented on plugins/usepe_logger.py:
```diff
@@ -63,3 +74,34 @@ def update(self):
+        currentlos = list()
+
+        # Go through all loss of separation pairs and sort the IDs for easier matching
+        currentlos = [list(pair) for pair in traf.cd.lospairs_unique]
+        for pair in currentlos: pair.sort()
```
Replacing `list()` with `sorted()` would get rid of the `for` loop I think:
```suggestion
        currentlos = [sorted(pair) for pair in traf.cd.lospairs_unique]
```

joakimrydberg replied:
That works!

dulli reviewed (approved):
I don't really have a strong preference regarding whether to have it in multiple logs vs a single file. The former makes it easier to differentiate the event types while the latter makes it easier to see at a glance which events may have led to other events. But that could also be accomplished (in a better way I guess) using an easy post-processing/visualization step on the multiple files.

mbaena-nommon commented:
The pull request seems OK. We agree with @dulli, it can be done in a post-processing step. At some point it may be interesting to have the traceability to know which conflicts end up in a loss of separation, to get insights from that.

joakimrydberg committed:
Moving sorting of LoS pairs into the list comprehension

Co-authored-by: Martin Dullweber <martin.dullweber@dlr.de>

joakimrydberg committed:
Add comments in usepe_logger.py

- Added descriptive comments
- Removed commented out code

joakimrydberg committed:
Move conflict logger to its own function

Applied list comprehension, as with LoS

joakimrydberg committed:
Implement command for usepe_logger.py

2 options for the command:
 - LIST: Lists all the available loggers
 - ON: Enables all the available loggers

joakimrydberg commented:
I have applied the same list comprehension for conflicts as for LoS.

I have also implemented a command `USEPELOGGER` which has 2 options:
- `LIST`: Lists all the available loggers
- `ON`: Enables all the available loggers

This means that rather than enabling each logger individually in your scenario files, all you need is to invoke:
`USEPELOGGER ON`

As there are no conflicts, I will merge the pull request.
